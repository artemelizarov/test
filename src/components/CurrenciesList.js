import React, { Component } from "react";
import Currency from "./Currency";
import CurrencySelect from "./CurrencySelect";

class CurrenciesList extends Component {
  baseCurrencyOnChange = event => {
    const value = event.target.value;
    this.props.changeBaseCurrency(value);
  };

  componentDidMount() {
    this.props.changeBaseCurrency(this.props.defaultBaseCurrency);
  }

  render() {
    const {
      baseCurrency,
      errorRates,
      currenciesRates,
      currencies,
      changeFavorite
    } = this.props;

    return (
      <div className="CurrenciesList">
        <h2>Base currency</h2>
        <CurrencySelect
          label="Base currency"
          name="baseCurrency"
          currencies={currencies}
          currency={baseCurrency}
          onChange={this.baseCurrencyOnChange}
        />
        <h2>List of currencies</h2>
        {errorRates && <div>Error: {errorRates.message}</div>}
        <ul>
          {currenciesRates.map(currency => (
            <Currency
              key={currency.code}
              code={currency.code}
              name={currency.name}
              rate={currency.rate}
              favorite={currency.favorite}
              onChangeFavorite={changeFavorite}
            />
          ))}
        </ul>
      </div>
    );
  }
}

export default CurrenciesList;
