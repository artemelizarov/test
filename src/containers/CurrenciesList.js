import { connect } from "react-redux";
import { changeBaseCurrency, changeFavorite } from "../redux/actions";
import CurrenciesList from "../components/CurrenciesList";

const mapStateToProps = state => {
  let currenciesWithRates = [...state.currencies];
  const { currenciesRates, baseCurrency, favorites } = state.currenciesList;
  currenciesWithRates = currenciesWithRates.filter(
    currency => currency.code !== baseCurrency
  );
  if (currenciesRates) {
    currenciesWithRates.forEach(currency => {
      currency.rate = currenciesRates[currency.code];
      currency.favorite = favorites.includes(currency.code);
    });
  }
  currenciesWithRates.sort(a => (a.favorite ? -1 : 1));

  return {
    baseCurrency: state.currenciesList.baseCurrency,
    currenciesRates: currenciesWithRates,
    errorRates: state.currenciesList.errorRates,
    currencies: state.currencies,
    defaultBaseCurrency: state.defaultCurrency,
    favorites: state.favorites
  };
};

const mapDispatchToProps = dispatch => ({
  changeBaseCurrency: baseCurrency =>
    dispatch(changeBaseCurrency(baseCurrency)),
  changeFavorite: currency => dispatch(changeFavorite(currency))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CurrenciesList);
