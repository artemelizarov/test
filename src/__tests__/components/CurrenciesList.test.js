import CurrenciesList from "../../components/CurrenciesList";

describe("render CurrenciesList component", () => {
  let wrapper, _changeBaseCurrency;

  beforeAll(() => {
    _changeBaseCurrency = jest.fn();

    wrapper = mount(
      <CurrenciesList
        defaultBaseCurrency="USD"
        currencies={[
          { code: "USD", name: "United States dollar" },
          { code: "EUR", name: "Euro" }
        ]}
        currenciesRates={[
          { code: "USD", name: "United States dollar", rate: 1 },
          { code: "EUR", name: "Euro", rate: 2 }
        ]}
        changeBaseCurrency={_changeBaseCurrency}
      />
    );

    let input;
    input = wrapper.find("select").first();
    input.instance().value = "EUR";
    input.simulate("change");
  });

  test("CurrenciesList render", () => {
    expect(wrapper.find(".CurrenciesList").length).toBe(1);
  });

  test("_changeBaseCurrency is called", () => {
    expect(_changeBaseCurrency).toBeCalledWith("EUR");
  });
});

describe("snapshot-test CurrenciesList component", () => {
  test("Renders correct properties", () => {
    shallowExpect(
      <CurrenciesList
        defaultBaseCurrency="USD"
        currencies={[
          { code: "USD", name: "United States dollar" },
          { code: "EUR", name: "Euro" }
        ]}
        currenciesRates={[
          { code: "USD", name: "United States dollar", rate: 1 },
          { code: "EUR", name: "Euro", rate: 2 }
        ]}
        changeBaseCurrency={jest.fn()}
      />
    ).toMatchSnapshot();
  });
});
