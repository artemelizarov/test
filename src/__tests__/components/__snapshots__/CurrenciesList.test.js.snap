// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot-test CurrenciesList component Renders correct properties 1`] = `
<div
  className="CurrenciesList"
>
  <h2>
    Base currency
  </h2>
  <CurrencySelect
    currencies={
      Array [
        Object {
          "code": "USD",
          "name": "United States dollar",
        },
        Object {
          "code": "EUR",
          "name": "Euro",
        },
      ]
    }
    label="Base currency"
    name="baseCurrency"
    onChange={[Function]}
  />
  <h2>
    List of currencies
  </h2>
  <ul>
    <Currency
      code="USD"
      key="USD"
      name="United States dollar"
      rate={1}
    />
    <Currency
      code="EUR"
      key="EUR"
      name="Euro"
      rate={2}
    />
  </ul>
</div>
`;
